src/backend/distributed/executor/multi_task_tracker_executor.c

```diff
@@ -2449,66 +2455,116 @@ NextRunningTaskIndex(List *assignedTaskList, int32 currentTaskIndex)
 
 		if (taskRunning && (assignedTaskIndex > currentTaskIndex))
 		{
-			nextTaskIndex = assignedTaskIndex;
-			break;
+			taskStatusBatchList = lappend(taskStatusBatchList, assignedTask);
+			if (list_length(taskStatusBatchList) >= MaxTaskStatusBatchSize)
+			{
+				break;
+			}
 		}
 
 		assignedTaskIndex++;
 	}
 
-	/*
-	 * We might have wrapped around the assigned task list and found nothing. If
-	 * that's the case, we'll start looking from the beginning next time.
-	 */
-	return nextTaskIndex;
+	/* continue where we left off next time this function is called */
+	taskTracker->currentTaskIndex = assignedTaskIndex;
+
+	return taskStatusBatchList;
 }
 
 
 /*
- * TaskStatusQueryResponse assumes that a task status query has been previously
- * sent on the given connection, and reads the response for this status query.
+ * TaskStatusBatchQuery builds a command string containing multiple
+ * task_tracker_task_status queries from a TrackerTaskState list.
  */
-static TaskStatus
-TaskStatusQueryResponse(int32 connectionId)
+static StringInfo
+TaskStatusBatchQuery(List *taskList)
 {
-	TaskStatus taskStatus = TASK_STATUS_INVALID_FIRST;
-	void *queryResult = NULL;
+	StringInfo taskStatusBatchQuery = makeStringInfo();
+	ListCell *taskCell = NULL;
+
+	foreach(taskCell, taskList)
+	{
+		TrackerTaskState *taskState = (TrackerTaskState *) lfirst(taskCell);
+
+		if (taskStatusBatchQuery->len > 0)
```

Comment:
I think we don't need this IF - we can have a ; at the end. We can add ; to TASK_STATUS_QUERY, or if you want it to be exclusive, we can add it directly after taskAssignmentQuery->data.

I noticed that this is coming from the old code, but I think while updating here we can change it.